Gate Aggregate Performance Tests; Issue #158 (#165)
PennyLaneAI/pennylane-lightning: examples/README.md

Run `bash run_gate_benchmark.sh $CXX_COMPILER`, where `$CXX_COMPILER` is the compiler you wish to use, in the terminal (e.g. `bash run_gate_benchmark.sh clang++`). The script will automatically build the gate_benchmark project.
It will set the CXX environment variable to "$CXX_COMPILER".

Some comments on the implementation:
General comments: 
* The compile-time options are controlled by the bash script `run_gate_benchmark.sh`
* The lightning code is run in the cpp file

> * Plotting is done with the python script `gate_benchmark_plotter.py`

Reviewer: ```suggestion
* Plotting is accomplished with the Python script `gate_benchmark_plotter.py`. 
* Plotting requires the packages listed in `requirements.txt`
```

Be sure to add a `requirements.txt` file for the dependent Python packages to enable plotting. Looking at the Python file, this will need to include `numpy`, `matplotlib` and `pandas`.

> * The number of gate repitions is set to 3 and can be changed in the bash script `run_gate_benchmark.sh` by changing the `num_gate_reps` variable

Reviewer: ```suggestion
* The number of gate repetitions is set to 3 and can be changed in the bash script `run_gate_benchmark.sh` by modifying the `num_gate_reps` variable
```

> gate_benchmark.cpp:

Reviewer: ```suggestion
### `gate_benchmark.cpp`:
```

* A single random angle is generated per gate repetition and qubit; the same random angle is used once for all of the parameterised gates
* The gates are applied in the order X, Y, Z, H, CNOT, CZ, RX, RY, RZ, CRX, CRY, CRZ
* The above order is repeated `num_gate_reps`-times

> gate_benchmark_plotter.py:

Reviewer: ```suggestion
### `gate_benchmark_plotter.py`:
```

If the backticks renders the page strangely, feel free to ignore them.

* The first plot shows the absolute runtime
* The second plot is on a loglog scale which better depicts the exponential scaling of the relative runtime with respect to the number of simulated qubits

> * We plot the time needed to execute the gate sequence averaged over the repititions

Reviewer: ```suggestion
* We plot the time needed to execute the gate sequence averaged over the repetitions
```

> Run "bash run_gate_benchmark.sh" in the terminal. The script will automatically build the gate_benchmark project.

Reviewer: Using backticks here renders the output as code (single for inline, triple for blocks).

> Run "bash run_gate_benchmark.sh" in the terminal. The script will automatically build the gate_benchmark project.

Author: Incorporated your changes. Thanks for the hint!

> It will set the CXX and CMAKE_CXX_FLAGS environment variables to "which clang++" and "-O3".

Reviewer: It would be great if this were user-selectable, rather than set in the script (even as a command-line argument). Other potential options could be `g++`, `icpc`, `icpx`, `nvc++` and others.

Author: The compiler name now has to be specified as a command-line argument after `run_gate_benchmark.sh` script.

* I assume clang is installed
* I build the project in the bash script to ensure that the compiler and flags are consistent with the ones that appear in the plot description
* The number of gate repitions was set to 3 as described in the email

gate_benchmark.cpp:
* I decided to generate a single random angle per parameterised gate as I feel like this makes use of parametric property of the gates while keeping the performance penalty due to cache access to a bare minimum
* I decided to apply the gates in the order that they were listed in the task; furthermore I perform the gate repitition after every single gate has been applied to avoid performing identity operations for gates which are involutory
* I have included many different libraries without much regard as to which libraries are already used in pennylane-lightning as I don't believe this is the focus of this task

gate_benchmark_plotter.py:
* The first plot is supposed to be close to the task description
* Due to the exponential nature of quantum computing one cannot read out the time for few qubits; I therefore also included a loglog plot depicting the scaling of the program

> * Only the average is plotted as described in the task description. Normally I would average over many more runs and plot the 95% confidence interval or standard deviation

Reviewer: Great job on the description here. Though, one thing to note is this readme could server as a way to document the example benchmark of the library using your implemented work. In this case, I think the phrasing needs to focus more on what is happening as is, with less focus on your decisions.

eg
*I decided to only run the lightning code in the cpp file and to control the compilation, file-creation etc via a bash script*
could be replaced with

*The compile-time options are controlled by the bash script `scriptname.sh`

Author: Changed the `README.md` to be more concise and to actually describe what was implemented. The first version was meant to clear up any questions regarding my thought-process in the context of the coding challenge.
